llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/TaskExecutorService.java

```diff
@@ -884,10 +885,20 @@ private void finishableStateUpdated(TaskWrapper taskWrapper, boolean newFinishab
         taskWrapper.updateCanFinishForPriority(newFinishableState);
         forceReinsertIntoQueue(taskWrapper, isRemoved);
       } else {
-        taskWrapper.updateCanFinishForPriority(newFinishableState);
-        if (!newFinishableState && !taskWrapper.isInPreemptionQueue()) {
-          // No need to check guaranteed here; if it was false we would already be in the queue.
+        // if speculative task, any finishable state change should re-order the queue as speculative tasks are always
+        // not-guaranteed (re-order helps put non-finishable's ahead of finishable)
+        if (!taskWrapper.isGuaranteed()) {
+          removeFromPreemptionQueue(taskWrapper);
+          taskWrapper.updateCanFinishForPriority(newFinishableState);
           addToPreemptionQueue(taskWrapper);
+        } else {
+          // if guaranteed task, if the finishable state changed to non-finishable and if the task doesn't exist
+          // pre-emption queue, then add it so that it becomes candidate to kill
+          taskWrapper.updateCanFinishForPriority(newFinishableState);
```

Comment:
Can there be a case where we have a Guaranteed task that changes from non-finishable to finishable and is only part of the preemptionQueue?
Under that scenario our code (and the old code) would remove it from preemptionQ and it would not be part of any other Q.

From the code below it seems that this can indeed happen:
https://github.com/apache/hive/blob/master/llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/TaskExecutorService.java#L776

Reply:
Non-finishable -> Finishable does not have to be pre-emption queue. This could be in wait queue (if not capacity) or taken by executor to run both which are fine.
You brought up good point, we may be adding the same task fragment to pre-emption queue twice. I will add a if not exists check when adding to pre-emption queue.

Reply:
@prasanthj thanks for fixing this! Patch looks good and is now committed!
At some point I would also change the first comment of the method to clarify that a task that is both Guaranteed and Finishable should never be in the preemption queue. 
https://github.com/apache/hive/pull/1012/files#diff-16658bf15468ecd089c4fd32e75fa8b2R876

I am believe there is value to put some effort documenting and describing how the scheduler works (started keeping some noted but happy to add more work). I personally find the information in this area of the project very limited.